Scaling up with reducer and context ru (#981)
reactjs/ru.react.dev: src/content/learn/scaling-up-with-reducer-and-context.md

---
title: Масштабирование с помощью редюсера и контекста
---

<Intro>

Редюсеры позволяют объединить и упорядочить логику обновления состояния. Контекст позволяет передавать состояние глубоко в дочерние компоненты. Вы можете объединить редюсеры и контекст для управления комплексом состояний.

</Intro>

<YouWillLearn>

* Как объединить редюсер и контекст
* Как избежать передачи состояния и отправителя через пропсы
* Как хранить логику контекста и состояния в отдельном файле

</YouWillLearn>

## Объединение редюсера с контекстом {/*combining-a-reducer-with-context*/}

> В этом примере из [введения в редюсеры](/learn/extracting-state-logic-into-a-reducer), состояние контролируется редюсером. Функция редюсер содержит в себе всю логику обновления состояния и объявлена в нижней части этого файла:

Reviewer: ```suggestion
В этом примере из [введения в редюсеры](/learn/extracting-state-logic-into-a-reducer), состояние контролируется редюсером. Функция-редюсер содержит в себе всю логику обновления состояния и объявлена в нижней части этого файла:
```
В данном случае это сложное существительное, состоящее из двух слов, для которого нужно использовать дефис

</Sandpack>

> Редюсер помогает сделать обработчики событий краткими и лаконичными. Однако по мере роста вашего приложения вы можете столкнуться с другой трудностью. **В настоящее время, состояние `tasks` и функция `dispatch` доступны только на верхнем уровне, в компоненте `TaskApp`.** Что бы дать доступ к чтению и изменению списка задач другим компонентам, вам надо явно  [передать в дочернии компоненты](/learn/passing-props-to-a-component) в виде пропсов, текущее состояние и обработчики событий для его изменения.

Reviewer: ```suggestion
Редюсер помогает сделать обработчики событий краткими и лаконичными. Однако по мере роста вашего приложения вы можете столкнуться с другой трудностью. **В настоящее время, состояние `tasks` и функция `dispatch` доступны только на верхнем уровне, в компоненте `TaskApp`.** Чтобы дать доступ к чтению и изменению списка задач другим компонентам, вам надо явно  [передать в дочерние компоненты](/learn/passing-props-to-a-component) в виде пропсов, текущее состояние и обработчики событий для его изменения.
```

> Например, `TaskApp` передает список задач и обработчики событий в `TaskList`:

Reviewer: ```suggestion
Например, `TaskApp` передаёт список задач и обработчики событий в `TaskList`:
```
пожалуйста, проверьте использование буквы ё с помощью пакета [eyo](https://github.com/e2yo/eyo)

/>
```

Это небольшой пример который хорошо работает, но если у вас десятки или сотни компонентов между начальной точкой, где создаются состояние и функции, и конечной точкой, где они будут использованы, передача их через пропсы может быть крайне затруднительной!

> И поэтому в качестве альтернативы передачи их через пропсы, вы можете поместить в [контекст](/learn/passing-data-deeply-with-context) состояние `tasks` и функцию `dispatch` **Благодаря этому дочернии компоненты `TaskApp` получить доступ к `tasks` и отправлять действия без передачи пропсов через компоненты которым они не нужны.**

Reviewer: ```suggestion
И поэтому в качестве альтернативы передачи их через пропсы, вы можете поместить в [контекст](/learn/passing-data-deeply-with-context) состояние `tasks` и функцию `dispatch` **Благодаря этому дочерние компоненты `TaskApp` смогут получить доступ к `tasks` и отправлять действия без передачи пропсов через компоненты, которым они не нужны.**
```

> Вот как вы можешь объединить редюсер с контекстом:

Reviewer: ```suggestion
Вот как вы можете объединить редюсер с контекстом:
```

1. **Создайте** контекст.
2. **Поместите** состояние и функцию отправитель в контекст.
3. **Используйте** контекст в любых дочерних элементах.

### Шаг 1: Создайте контекст {/*step-1-create-the-context*/}

> Хук `useReducer` возвращает текущее состояние `tasks` и функцию `dispatch` которая позволяет его обновлять:

Reviewer: ```suggestion
Хук `useReducer` возвращает текущее состояние `tasks` и функцию `dispatch`, которая позволяет его обновлять:
```

```js
const [tasks, dispatch] = useReducer(tasksReducer, initialTasks);
```

> Что бы передать их дальше по дереву вам надо [создать](/learn/passing-data-deeply-with-context#step-2-use-the-context) два раздельных контекста:

Reviewer: ```suggestion
Чтобы передать их дальше по дереву вам надо [создать](/learn/passing-data-deeply-with-context#step-2-use-the-context) два раздельных контекста:
```

- `TasksContext` содержит текущий список задач.

> - `TasksDispatchContext` содержит функции которая позволяет компонентам отправлять действия(`dispatch`).

Reviewer: ```suggestion
- `TasksDispatchContext` содержит функцию, которая позволяет компонентам отправлять действия(`dispatch`).
```

> Экспортируйте их из разных файлов что бы потом импортировать в других файлах:

Reviewer: ```suggestion
Экспортируйте их из разных файлов, чтобы потом импортировать в других файлах:
```
Повторяющаяся ошибка. Союз "чтобы" пишется слитно - проверьте по тексту, я мог где-то проглядеть.

</Sandpack>

В этом примере вы можете передать `null` как значение по умолчанию в каждый контекст. Актуальное значение передаст компонент `TaskApp`.

### Шаг 2: Поместите состояние и отправителя в контекст {/*step-2-put-state-and-dispatch-into-context*/}

> Сейчас вы можете импортировать контексты которые создавали в компонент `TaskApp`. Поместите `tasks` и `dispatch` возвращенные хуком `useReducer()` и [передайте их](/learn/passing-data-deeply-with-context#step-3-provide-the-context) во все дочерние компоненты:

Reviewer: ```suggestion
Сейчас вы можете импортировать контексты, которые создавали в компонент `TaskApp`. Поместите `tasks` и `dispatch`, возвращенные хуком `useReducer()`, и [передайте их](/learn/passing-data-deeply-with-context#step-3-provide-the-context) во все дочерние компоненты:
```

</Sandpack>

> На следующем шаге, вы избавитесь от пропсов.

Reviewer: ```suggestion
На следующем шаге вы избавитесь от пропсов.
```
а вот тут лишняя запятая, в русском такая конструкция не обособляется

### Шаг 3: Используйте контекст в любом месте дерева {/*step-3-use-context-anywhere-in-the-tree*/}

> Теперь вам не надо передавать список задач или слушатели событий в дочернии компоненты:

Reviewer: ```suggestion
Теперь вам не надо передавать список задач или слушатели событий в дочерние компоненты:
```

```js {4-5}
<TasksContext.Provider value={tasks}>
  <TasksDispatchContext.Provider value={dispatch}>
    <h1>День в Киото</h1>
    <AddTask />
    <TaskList />
  </TasksDispatchContext.Provider>
</TasksContext.Provider>
```

> Вместо этого, все компоненты которым нужен список задач, могут получить его из `TaskContext`:

Reviewer: ```suggestion
Вместо этого все компоненты, которым нужен список задач, могут получить его из `TaskContext`:
```

```js {2}
export default function TaskList() {
  const tasks = useContext(TasksContext);
  // ...
```

> Что бы обновить список задач, любой компонент может получить функцию `dispatch` из контекста и использовать ее:

Reviewer: ```suggestion
Чтобы обновить список задач, любой компонент может получить функцию `dispatch` из контекста и использовать её:
```

export const TasksDispatchContext = createContext(null);
```

Этот файл скоро будет переполнен! Вам надо переместить редюсер в этот же файл. Затем создать новый компонент `TasksProvider` в этом же файле. Этот компонент соберет все части воедино:

1. Он будет управлять состоянием с помощью редюсера.
2. Он будет передавать оба контекста дочерним компонентам.

> 3. Он будет [принимать `children` как пропс](/learn/passing-props-to-a-component#passing-jsx-as-children) что бы вы могли передавать JSX в него.

Reviewer: ```suggestion
3. Он будет [принимать `children` как проп,](/learn/passing-props-to-a-component#passing-jsx-as-children) чтобы вы могли передавать JSX в него.
```

</Sandpack>

> Вы так—же можете экспортировать функции которые _используют_ контекст из `TasksContext.js`:

Reviewer: ```suggestion
Вы также можете экспортировать функции, которые _используют_ контекст из `TasksContext.js`:
```
Союз "также" (в значении и, и ещё) пишется слитно

</Sandpack>

Можно представить TasksProvider как часть интерфейса, которая знает, как работать с задачами, useTasks — как способ их чтения, а useTasksDispatch — как способ обновления из любого дочернего компонента.

<Note>

> Функции похожие на `useTasks` и `useTasksDispatch` называются *[Пользовательские  хуки.](/learn/reusing-logic-with-custom-hooks)* Ваша функция считается пользовательским хуком если ее название начинается с `use`. Это позволяет вам использовать другие хуки, такие как `useContext`, внутри неё.

Reviewer: ```suggestion
Функции, похожие на `useTasks` и `useTasksDispatch`, называются *[Пользовательские  хуки.](/learn/reusing-logic-with-custom-hooks)* Ваша функция считается пользовательским хуком, если её название начинается с `use`. Это позволяет вам использовать другие хуки, такие как `useContext`, внутри неё.
```

</Note>

По мере роста вашего приложения у вас может появиться много таких контекст-редюсер пар. Это мощный способ масштабировать приложение и [передавать состояние](/learn/sharing-state-between-components) без лишних усилий, когда вам нужно получить доступ к данным глубоко в дереве.

<Recap>

> - Вы можете комбинировать редюсер с контекстом, что бы любой компонент мог получить доступ к состоянию родительского компонента и изменять его

Reviewer: ```suggestion
- Вы можете комбинировать редюсер с контекстом, чтобы любой компонент мог получить доступ к состоянию родительского компонента и изменять его
```

- Чтобы предоставить состояние и функцию отправителя в дочернии компоненты, надо:
  1. Создать два контекста (для состояния и функций отправителей).

>   2. Экспортируйте оба контекста из компонента в котором объявлен редюсер.

Reviewer: ```suggestion
  2. Экспортируйте оба контекста из компонента, в котором объявлен редюсер.
```
